// ship/lib/deployers/bitballoon/index.js
// Generated by CoffeeScript 1.9.1
(function() {
  var W, _, bitballoon, create, deploy, destroy, lookup, node;

  bitballoon = require('bitballoon');

  W = require('when');

  node = require('when/node');

  _ = require('lodash');

  module.exports = function(root, config) {
    var client, d;
    d = W.defer();
    if (!config.access_token) {
      return d.reject('missing access_token!');
    }
    client = bitballoon.createClient({
      access_token: config.access_token
    });
    W()["with"]({
      root: root,
      client: client,
      config: config,
      d: d
    }).then(lookup).then(function(site) {
      if (site) {
        return site;
      } else {
        return create.call(this);
      }
    }).then(deploy).done(function(site) {
      return d.resolve({
        deployer: 'bitballoon',
        url: site.url,
        destroy: destroy.bind(this, site.site_id)
      });
    }, d.reject);
    return d.promise;
  };


  /**
   * Checks to see if your site is already on bitballoon or not. Returns either a
   * site object or undefined.
   * @return {Promise} promise for either undefined or a site object
   */

  lookup = function() {
    return node.call(this.client.sites.bind(this.client)).then((function(_this) {
      return function(sites) {
        return _.find(sites, {
          name: _this.config.name
        });
      };
    })(this));
  };


  /**
   * Creates a new site on bitballoon with a given name.
   * @return {Promise} promise for a newly created site object
   */

  create = function() {
    this.d.notify("Creating '" + this.config.name + "' on bitballoon");
    return node.call(this.client.createSite.bind(this.client), {
      name: this.config.name
    });
  };


  /**
   * Creates a new deploy for a given site with the contents of the root.
   * @param {Object} site - a bit object from bitballoon
   * @return {Promise} a promise for a finished deployment
   */

  deploy = function(site) {
    this.d.notify("Deploying '" + this.config.name + "'");
    return node.call(site.createDeploy.bind(site), {
      dir: this.root
    }).then(function(deploy) {
      return node.call(deploy.waitForReady.bind(deploy));
    });
  };


  /**
   * Deletes a given site from bitballoon.
   * @param  {Object} site - a site object from bitballoon
   * @return {Promise} a promise for the deleted site
   */

  destroy = function(id) {
    return node.call(this.client.site.bind(this.client), id).then(function(site) {
      return node.call(site.destroy.bind(site));
    });
  };

}).call(this);
