// roots/lib/api/deploy.js
// Generated by CoffeeScript 1.9.1
(function() {
  var Config, Deploy, Ship, W, fs, path;

  W = require('when');

  fs = require('fs');

  path = require('path');

  Ship = require('ship');

  Config = require('../config');

  Deploy = (function() {
    function Deploy(project) {
      this.project = project;
    }

    Deploy.prototype.exec = function(opts) {
      if (opts == null) {
        opts = {};
      }
      __track('api', {
        name: 'deploy',
        deployer: opts.to
      });
      return this.project.clean().then((function(_this) {
        return function() {
          return _this.project.compile();
        };
      })(this)).then((function(_this) {
        return function() {
          if (fs.existsSync(path.join(_this.project.root, 'ship.production.conf'))) {
            return new Ship({
              root: _this.project.root,
              deployer: opts.to,
              env: 'production'
            });
          } else {
            return new Ship({
              root: _this.project.root,
              deployer: opts.to
            });
          }
        };
      })(this)).tap(function(ship) {
        if (!ship.is_configured()) {
          return ship.config_prompt().tap(function(values) {
            return ship.config = values;
          }).then(function() {
            return ship.write_config();
          });
        }
      }).tap((function(_this) {
        return function(ship) {
          return ship.deploy(_this.project.config.output_path());
        };
      })(this));
    };

    return Deploy;

  })();

  module.exports = Deploy;

}).call(this);
